src/package.rs

```diff
@@ -158,21 +85,58 @@ fn add_directory_and_parents<W: std::io::Write>(
     Ok(())
 }
 
-/// Trait for propagating progress information while constructing the package.
-pub trait Progress {
-    /// Updates the message displayed regarding progress constructing
-    /// the package.
-    fn set_message(&self, msg: impl Into<Cow<'static, str>>);
+/// Describes the origin of an externally-built package.
+#[derive(Deserialize, Debug)]
+#[serde(tag = "type", rename_all = "lowercase")]
+pub enum PackageSource {
```

Comment:
This is the crux of this change: `PackageSource` and `PackageOutput` describe the new package types.